Feature/itemeffects and enchants

InDebt opened the pull request:
* Adds the a set of auto generated on use and proc effects
* Right now only stat effects are supported
* Every effect can be overwritten by a specific implementation
* I have added unimplementeded enchants to the list of enchants even when they're not stat effects to make sure they're visible

NerdEgghead commented on tools/database/gen_effects.go:
```diff
@@ -0,0 +1,591 @@
+
+	return strings.Join(callbacks, " | ")
+}
+
+func asCoreProcMask(procMask core.ProcMask) string {
```
Rather than hardcoding all of this manually, it would be cleaner and more maintainable to instead implement `String()` methods on the `AuraCallback`, `ProcMask`, and `HitOutcome` types within `core`. These methods can be generated automatically using Go's `generate` feature:
https://pkg.go.dev/golang.org/x/tools/cmd/stringer

If you have `String()` methods for each individual component, then you could write the below code more generally:
```
func (pm ProcMask) GetComponentFlags() []ProcMask {
    var components []ProcMask
    for i := 0; i < 32; i++ {
        componentFlag := ProcMask(1 << i)
    
        if componentFlag >= ProcMaskLast { // Add ProcMaskLast to the end of the enum
            break
        }
        
        if pm.Matches(componentFlag) {
            components = append(components, componentFlag)
        }
    }
    
    return components
}

func (pm ProcMask) GenerateGoCode() string {
    components := pm.GetComponentFlags()
    componentStrings := MapSlice(components, func(componentFlag ProcMask) string {
        return "core." + componentFlag.String()
    })
    return strings.Join(componentStrings, " | ")
}
```

InDebt replied:
Did this now for AuraCallback and ProcMask. stringer generates the fiels int he module.. should we git ignore them and add them to the make prot or any other comon action? Right now I've checked them in.

NerdEgghead commented on tools/database/gen_effects.go:
```diff
@@ -0,0 +1,591 @@
+
+	return false
+}
+
+func formatStrings(maxLength int, input []string) []string {
```
This should live in a util file.

NerdEgghead commented on tools/tooltip/tooltip_parser.go:
```diff
@@ -1015,7 +1015,7 @@ func ParseTooltip(tooltip string, dataProvider TooltipDataProvider, spellId int6
 		"PL":           dataProvider.GetPlayerLevel(),
 		"pl":           dataProvider.GetPlayerLevel(),
-		"proccooldown": dataProvider.GetSpellProcChance(spellId),
+		"proccooldown": math.Round(float64(dataProvider.GetSpellProcCooldown(spellId))/float64(time.Second)*100) / 100,
```
This can be simplified by calling `dataProvider.GetSpellProcCooldown(spellId).Seconds()` instead.

InDebt commented on sim/common/mop/enchants_auto_gen.go:
```diff
@@ -0,0 +1,208 @@
+		Outcome:   core.OutcomeEmpty,
+	})
+	
+	// NOTE: Manual implementation required
+	//       This can be ignored if the effect has already been implemented.
+	//
+	// Permanently enchants a melee weapon to sometimes increase your dodge by 1650 for 7s when dealing melee
+	// damage.
+	shared.NewProcStatBonusEffect(shared.ProcStatBonusEffect{
+		Name:     "Enchant Weapon - River's Song",
+		EnchantID: 4446,
+		Callback:  core.CallbackEmpty,
+		ProcMask:  core.ProcMaskEmpty,
+		Outcome:   core.OutcomeEmpty,
+	})
```
https://wago.tools/db2/SpellEffect?build=5.5.0.61208&filter%5BSpellID%5D=116660%7C104441&page=1
https://wago.tools/db2/SpellItemEnchantment?build=5.5.0.61208&filter%5BID%5D=4446&page=1

For some reason only the effect arg 1 is used and not the stat buff. Either the enchant record is broken or the dummy uses the enchant arg 0 to buff the effect and there is some scrpting going on for this enchant

InDebt commented on sim/common/mop/enchants_auto_gen.go:
```diff
@@ -0,0 +1,208 @@
+
+	// Enchants
+	
+	// NOTE: Manual implementation required
+	//       This can be ignored if the effect has already been implemented.
+	//
```
My first reaction was - no.

However - we're importing the core module, so we should be able to include the common package as well, that registers all functions in the init call - so I should have access to the ItemEffect lookups 🤔  I'll take a look

InDebt replied:
Oh yeah there's no note the trinkets are just removed from the auto gen code if they're already present 😁

InDebt commented on sim/common/mop/stat_bonus_procs_augo_gen.go:
```diff
@@ -0,0 +1,2768 @@
+		Outcome:  core.OutcomeLanded,
+	})
+	
+	// [Each time your spells heal you have a chance to gain 963 Intellect for 20s. ( 15% chance, 55 sec cooldown)]
+	shared.NewProcStatBonusEffect(shared.ProcStatBonusEffect{
+		Name:     "Qin-xi's Polarizing Seal",
+		ItemID:   86805,
+		Callback: core.CallbackOnHealDealt | core.CallbackOnPeriodicDamageDealt | core.CallbackOnPeriodicHealDealt | core.CallbackOnSpellHitDealt,
+		ProcMask: core.ProcMaskSpellDamage | core.ProcMaskSpellDamageProc,
+		Outcome:  core.OutcomeLanded,
+	})
```
Okay this one went through.

The proc masks are actually included in the dbc. They would not work for our sim though, but it's hard to detect 'only on heal trinkets' - i'll check it

InDebt commented on sim/common/mop/stat_bonus_procs_augo_gen.go:
```diff
@@ -0,0 +1,2768 @@
+	
+	// NOTE: Manual implementation required
+	//       This can be ignored if the effect has already been implemented.
+	// [When you dodge, you have a 4% chance to gain 963 mastery for 20s. This effect can stack up to 3 times.]
```
👍 Done

InDebt commented on tools/database/gen_effects.go:
```diff
@@ -0,0 +1,591 @@
+func asCoreCallback(callback core.AuraCallback) string {
+	callbacks := []string{}
+	if callback.Matches(core.CallbackOnApplyEffects) {
+		callbacks = append(callbacks, "core.CallabckOnApplyEffects")
```
Should be resolved with refactor to generic

NerdEgghead commented:
> Alright I did some passes today. Will go at it again tomorrow, addressing the missing scalign for some effect types and cleanup some procs / kick out some spells.

Take your time, just tag me for a re-review whenever you feel it is appropriate. All this core work is immensely appreciated!

InDebt commented on sim/common/shared/shared_utils.go:
```diff
@@ -51,7 +46,7 @@ func NewProcStatBonusEffectWithDamageProc(config ProcStatBonusEffect, damage Dam
 	}
 
-	factory_StatBonusEffect(config, func(agent core.Agent) ExtraSpellInfo {
+	factory_StatBonusEffect(config, func(agent core.Agent, _ proto.ItemLevelState) ExtraSpellInfo {
```
There are som DoT trinkets that I could find quickly but we're not supporting them right now.

NerdEgghead commented on tools/database/utils.go:
```diff
@@ -62,3 +65,135 @@ func ParseRandomSuffixOptions(optionsString sql.NullString) ([]int32, error) {
+
+func SpellUsesStacks(spellId int, instance *dbc.DBC) bool {
+	if spell, ok := instance.Spells[spellId]; ok {
+		if spell.MaxCumulativeStacks > 1 {
+			return true
+		}
```
Is the stat bonus per stack also contained in the DB? If so, could we generalize `shared.ProcStatBonusEffect` to handle both stacking and non-stacking buffs?

NerdEgghead replied:
Ah okay, if it's a mix then it might be better to stick with always defining those ones manually. But in that case we'll definitely need to have `shared.StackingStatBonusEffect` support scaling items.

NerdEgghead commented on sim/druid/guardian/TestGuardian.results:
Also, why aren't all the base settings test results changing as well? The test config has both Searing Words and Flashing Steel Talisman equipped, so I would expect DPS increases on all tests across the board once those trinkets are modeled with real effects on them.

InDebt commented on proto/common.proto:
```diff
@@ -133,6 +136,12 @@ enum Class {
 	ClassWarlock = 9;
 	ClassMonk = 10;
 	ClassDruid = 11;
+	ClassExtra1 = 12;
+	ClassExtra2 = 13;
+	ClassExtra3 = 14;
+	ClassExtra4 = 15;
+	ClassExtra5 = 16;
+	ClassExtra6 = 17;
```
But they refer to different scaling classes. So you want them to be called Uknown1 to 6?

InDebt commented on sim/common/mop/stat_bonus_procs_auto_gen.go:
```diff
@@ -0,0 +1,1649 @@
+	//	{ItemID: 105472, ItemName: "Assurance of Consequence (Heroic Warforged)"},
+	// })
+	
+	// TODO: Manual implementation required
+	//       This can be ignored if the effect has already been implemented.
+	//       With next db run the item will be removed if implemented.
+	//
+	// Your attacks have a chance to grant 1 Intellect for 20s. ( 15% chance, 115 sec cooldown)
```
Right now we only support 1 effect on an item.
This one has kinda two with this effect here: https://www.wowhead.com/mop-classic/spell=146051/amplification
Similar to the Crit Meta gems that need manual code right now. That is not supported right now

ToxicKevinFerm replied:
We weighed the risk of having more than one effect on an item against the db size in the Cata repo :D

NerdEgghead replied:
Yeah it's fine to just require a manual implementation in cases like this, there aren't too many of them. Just wanted to make sure it was a valid reason and not a potential bug, that's all.

InDebt commented on sim/common/mop/stat_bonus_procs_auto_gen.go:
```diff
@@ -0,0 +1,1649 @@
+		Outcome:  core.OutcomeLanded,
+		Harmful:  false,
+	}, []shared.ItemVariant{
+		{ItemID: 81133, ItemName: "Empty Fruit Barrel (N)"},
+		{ItemID: 97304, ItemName: "Empty Fruit Barrel"},
```
I think that's part of the db cleanup that Net is working on

InDebt commented on sim/common/mop/stat_bonus_procs_auto_gen.go:
```diff
@@ -0,0 +1,1649 @@
+		{ItemID: 97304, ItemName: "Empty Fruit Barrel"},
+	})
+	
+	// When your attacks hit you have a chance to gain 2573 critical strike for 30s. ( 15% chance, 115 sec cooldown)
+	shared.NewProcStatBonusEffect(shared.ProcStatBonusEffect{
+		Name:     "Carbonic Carbuncle (N)",
+		ItemID: 81138,
+		Callback: core.CallbackOnSpellHitDealt,
+		ProcMask: core.ProcMaskMeleeMHAuto | core.ProcMaskMeleeOHAuto | core.ProcMaskMeleeMHSpecial | core.ProcMaskMeleeOHSpecial | core.ProcMaskRangedAuto | core.ProcMaskRangedSpecial | core.ProcMaskSpellDamage | core.ProcMaskMeleeProc | core.ProcMaskRangedProc | core.ProcMaskSpellDamageProc,
+		Outcome:  core.OutcomeLanded,
+		Harmful:  false,
+	})
+	
+	// When your spells deal damage you have a chance to gain 2573 critical strike for 30s. ( 15% chance, 115
+	// sec cooldown)
+	shared.NewProcStatBonusEffect(shared.ProcStatBonusEffect{
+		Name:     "Vision of the Predator (N)",
+		ItemID: 81192,
+		Callback: core.CallbackOnSpellHitDealt | core.CallbackOnPeriodicDamageDealt,
+		ProcMask: core.ProcMaskSpellDamage | core.ProcMaskSpellDamageProc,
+		Outcome:  core.OutcomeLanded,
+		Harmful:  false,
+	})
```
Yes that's correct. I went on beta just now end tested. The melee one procs off pure Lightning Bolt spam

NerdEgghead replied:
Ok cool, just wanted to make sure.

InDebt commented on sim/common/mop/enchants_auto_gen.go:
```diff
@@ -0,0 +1,232 @@
+		Harmful:   false,
+	})
+	
+	// Embroiders a subtle pattern of light into your cloak, giving you a chance to increase your Intellect by
+	// 2000 for 15s when casting a spell.
+	// 
+	// Embroidering your cloak will cause it to become soulbound and requires the Tailoring profession to remain
+	// active.
+	shared.NewProcStatBonusEffect(shared.ProcStatBonusEffect{
+		Name:      "Lightweave Embroidery (Rank 3)",
+		EnchantID: 4892,
+		Callback:  core.CallbackOnSpellHitDealt | core.CallbackOnPeriodicDamageDealt | core.CallbackOnHealDealt | core.CallbackOnPeriodicHealDealt,
+		ProcMask:  core.ProcMaskSpellDamage | core.ProcMaskSpellHealing | core.ProcMaskSpellDamageProc,
+		Outcome:   core.OutcomeLanded,
+		Harmful:   false,
+	})
```
Actually not 😁 Can't believe you made me hit the training dummy for 45 min straight.

Soo.. 

1. It should be harmful
2. It should be Landed

It did not proc once off a miss and spamming i.E. Curse of Elements does also not proc it.
Looking at it, it might be saver to assume 'Harmful' for all spells and check if we can detect exceptions for non Hamful ones from the spells I've tested.

InDebt commented on sim/common/mop/enchants_auto_gen.go:
```diff
@@ -0,0 +1,232 @@
+	// Embroidering your cloak will cause it to become soulbound and requires the Tailoring profession to remain
+	// active.
+	shared.NewProcStatBonusEffect(shared.ProcStatBonusEffect{
+		Name:      "Lightweave Embroidery (Rank 3)",
+		EnchantID: 5110,
```
Well also junk spell here :( needs to be filtered out / blacklisted I guess.

InDebt commented on tools/database/dbc/item_effect.go:
```diff
@@ -33,3 +38,229 @@ func (e *ItemEffect) ToMap() map[string]interface{} {
+			}
+		}
+
+		// If proc chance is above 100 something weird is happening so we set ppm to 1 since we cant accurately proc it 100% of the time
+		if spTop.ProcChance == 0 || spTop.ProcChance > 100 {
+			proc.Ppm = 1
+			proc.ProcChance = 0
+		}
```
Yeah that's the thing I mentioned above. PPM values do not seem to be part of the dbc files. At least I could not make them out. So we somehow just default to 1 for all of them 😦 we now know it's most propably a PPM proc but we set wrong values here.

Not sure what's the best approach. Having a field LegacyPPM = true so we know it's suposed to be PPM?

InDebt commented on sim/common/mop/enchants_auto_gen.go:
```diff
@@ -0,0 +1,232 @@
+	//	Harmful:   false,
+	// })
+	
+	// Permanently attaches Lord Blastington's special scope to a ranged weapon, sometimes increasing Agility
+	// by 1800 for 10s when dealing damage with ranged attacks.
+	// 
+	// Attaching this scope to a ranged weapon causes it to become soulbound.
+	shared.NewProcStatBonusEffect(shared.ProcStatBonusEffect{
+		Name:      "Lord Blastington's Scope of Doom",
+		EnchantID: 4699,
+		Callback:  core.CallbackOnSpellHitDealt,
+		ProcMask:  core.ProcMaskRangedAuto | core.ProcMaskRangedSpecial,
+		Outcome:   core.OutcomeLanded,
+		Harmful:   false,
+	})
```
After more testing I think it's actually more accurate to assume harmful for all spells than not. I have not found a spell yet that procs on debuff apply for example what we would have as we deal the outcome with 0 dmg on apply.

InDebt replied:
Yeah, need to think about a good test place where we can test the behaviour and see if we can deduce it, if it's purely harmful or not and then how we distinguish that because non harmful spells would in the sim also proc on aura apply right now while ingame it might only spell that have a >0 pre mitigation value or so but we get to that then I guess :D

InDebt commented on sim/core/item_effects.go:
```diff
@@ -125,7 +125,7 @@ func (equipment *Equipment) applyItemEffects(agent Agent, registeredItemEffects
 // Helpers for making common types of active item effects.
 
 func NewSimpleStatItemActiveEffect(itemID int32, bonus stats.Stats, duration time.Duration, cooldown time.Duration, sharedCDFunc func(*Character) Cooldown, otherEffects ApplyEffect) {
-	NewItemEffect(itemID, func(agent Agent) {
+	NewItemEffect(itemID, func(agent Agent, _ proto.ItemLevelState) {
```
It can if we delete some cata trinkets. 

Moonwell Chalice and Shard of Woe use Helper Functions that use this method

InDebt commented on tools/database/gen_effects.go:
```diff
@@ -0,0 +1,642 @@
+	EffectParseResultSuccess     int = 2 // Returned when the effect was parsed successfuly
+)
+
+// Define your groups and effects here.
+// The map key is the group name, and the inner map is ID -> display name.
```
I think it was supposed to desribe how the groups param is structured 🤔 wasn't my comment removed as it's rather confusing

InDebt commented:
Alright I cleaned up a lot.

I need to get to some things still. For now please just check the shared_utils Stat Effect change if that is what you had in mind for the support.

To poperly filter ItemEffects I moved them down again and not merge the items in the effect generation, that caused us to properly apply our filters so now a lot of cata items were gone and the effects not loaded. So for any test to run I removed all old cata and wotlk effects we don't need in MOP.

So for now the biggest changes are:
* Item Filtered
* All Item Effects are Harmful

Topics for disucsion:
* Correct effect scaling. We need -1, -2 -4 for item effects and enchants. If you want to solve that differently than simply having them in the class enum how they're listed in the DBC / SpellScaling.txt we have to see how

NerdEgghead commented on sim/common/shared/shared_utils.go:
```diff
@@ -94,36 +110,47 @@ func factory_StatBonusEffect(config ProcStatBonusEffect, extraSpell func(agent c
+			item := core.GetItemByID(effectID)
+			if item.ItemEffect != nil {
+				if item.ItemEffect.GetProc() != nil {
+					procEffect = item.ItemEffect
+				}
+			}
 		}
 
-		procID := core.ActionID{SpellID: config.AuraID}
-		if procID.IsEmptyAction() {
-			procID = core.ActionID{ItemID: config.ItemID}
+		if procEffect == nil {
+			err, _ := fmt.Printf("Error getting proc effect for item/enchant %v", effectID)
+			panic(err)
 		}
-		procAura := character.NewTemporaryStatsAura(config.Name+" Proc", procID, config.Bonus, config.Duration)
+
+		proc := procEffect.GetProc()
+		procAction := core.ActionID{SpellID: procEffect.BuffId}
+		procAura := character.NewTemporaryStatsAura(
+			config.Name+" Proc",
+			procAction,
+			stats.FromProtoMap(procEffect.ScalingOptions[int32(itemLevelState)].Stats),
+			time.Millisecond*time.Duration(procEffect.EffectDurationMs),
+		)
 
 		var dpm *core.DynamicProcManager
-		if config.PPM != 0 {
-			if config.ProcMask == core.ProcMaskUnknown {
-				if isEnchant {
-					dpm = character.NewDynamicLegacyProcForEnchant(effectID, config.PPM, 0)
-				} else {
-					dpm = character.NewDynamicLegacyProcForWeapon(effectID, config.PPM, 0)
-				}
-			} else {
-				dpm = character.NewLegacyPPMManager(config.PPM, config.ProcMask)
-			}
+		if proc.Ppm != 0 {
+			dpm = character.NewRPPMProcManager(effectID, isEnchant, config.ProcMask, core.RppmConfigFromProcEffectProto(proc))
 		}
```
1 PPM looked to be about right for Windwalk and Gnomish X-Ray Scope based on in-game testing in Cata. You are correct that those values were just pure guesses (since 1.0 is the most common PPM value Blizzard uses in the old system) and not contained in the DB anywhere, but the guesses worked fine. Making the distinction would be good for generality though, we wouldn't want to accidentally parse a legacy PPM effect as an RPPM proc.

NerdEgghead reviewed (changes requested):
We also need to merge in:
https://github.com/wowsims/cata/pull/1498/files

InDebt commented on tools/database/gen_effects.go:
```diff
@@ -0,0 +1,637 @@
+		info.ProcMask |= core.ProcMaskSpellDamageProc
+	}
+
+	findOutcome := func() core.HitOutcome {
+		if critMatcher.MatchString(tooltip) {
+			return core.OutcomeCrit
+		}
+
+		return core.OutcomeLanded
+	}
+
+	info.Outcome = findOutcome()
```
Haha yeah seems I refactored the code over time so it's only used once! Good catch.
